server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1304,97 +1316,89 @@ private static IndexMetadata.State excludeState(IndicesOptions options) {
             return excludeState;
         }
 
-        private static Map<String, IndexAbstraction> matches(Context context, String expression) {
-            SortedMap<String, IndexAbstraction> indicesLookup = context.getState().getMetadata().getIndicesLookup();
-            if (Regex.isMatchAllPattern(expression)) {
-                return filterIndicesLookup(indicesLookup, null, context.getOptions().ignoreAliases(), context.includeDataStreams());
-            } else if (expression.indexOf("*") == expression.length() - 1) {
-                return suffixWildcard(indicesLookup, expression, context.getOptions().ignoreAliases(), context.includeDataStreams());
+        /**
+         * Given a single wildcard {@param expression}, return the {@code Stream} that contains all the resources (i.e. indices, aliases,
+         * and datastreams), that exist in the cluster at this moment in time, and that the wildcard "resolves" to (i.e. the resource's
+         * name matches the {@param expression} wildcard).
+         * The {@param context} provides the current time-snapshot view of all the resources, as well as conditions
+         * on whether to consider alias, datastream, system, and hidden resources.
+         * It does NOT consider the open or closed status of index resources.
+         */
+        private static Stream<IndexAbstraction> matchResourcesToWildcard(Context context, String wildcardExpression) {
```

Comment:
This method concentrates most (besides open/closed) resource wildcard matching and filtering in a stream-wise fashion.
Before, it was scattered accross two methods, separated by the `allow_no_indices` check.